### PR TITLE
[SPARK-33122][SQL][FOLLOWUP] Extend RemoveRedundantAggregates optimizer rule to apply to more cases

### PR DESCRIPTION
**Pull request by tanelk**
### What changes were proposed in this pull request?

Addressed the @dongjoon-hyun comments on the previous PR #30018.
Extended the `RemoveRedundantAggregates` rule to remove redundant aggregations in even more queries. For example in
 ```
dataset
    .dropDuplicates()
    .groupBy('a)
    .agg(max('b))
```
the `dropDuplicates` is not needed, because the result on `max` does not depend on duplicate values. 

### Why are the changes needed?

Improve performance.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    upperHasNoDuplicateSensitiveAgg && upperRefsOnlyDeterministicNonAgg
+  }
+
+  private def isAggregate(expr: Expression): Boolean = {
+    expr.find(e => e.isInstanceOf[AggregateExpression] ||
+      PythonUDF.isGroupedAggPandasUDF(e)).isDefined
+  }
+
+  private def isDuplicateSensitiveAggregate(expr: Expression): Boolean = {
```
Comment (tanelk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala:71]: Added this check, previously `isAggregate` was used twice.

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @tanelk .
Since the commit title is precious resource, please don't repeat the original JIRA title. The JIRA ID is enough for that purpose. It would be great if you give a more meaningful and specific PR title.

**Comment by tanelk** [comment]
@maropu , this is a followup to a PR you reviewed a while back, but it has gone unnoticed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      }
+  }
+
+  private def isLowerRedundant(upper: Aggregate, lower: Aggregate): Boolean = {
+    val upperHasNoDuplicateSensitiveAgg = upper
+      .aggregateExpressions
+      .forall(expr => expr.find {
+        case ae: AggregateExpression => !EliminateDistinct.isDuplicateAgnostic(ae.aggregateFunction)
+        case e => AggregateExpression.isAggregate(e)
+      }.isEmpty)
```
Comment (tanelk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala:58]: This is only behaviour change

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
I've checked the GA tests passed, so I will merge this. Thank you, @tanelk

**Comment by maropu** [comment]
Merged to master.

**Pull request closed without merging**